ethers-providers/src/provider.rs

```diff
@@ -2124,4 +2169,118 @@ mod tests {
+
+        let info = provider.node_info().await.unwrap();
+        drop(geth);
+
+        // check that the port we set works
+        assert_eq!(info.ports.listener, p2p_listener_port);
+
+        // make sure it is running eth
+        assert!(info.protocols.eth.is_some());
+
+        // check that the network id is correct
+        assert_eq!(info.protocols.eth.unwrap().network, network);
+    }
+
+    /// Spawn a new `GethInstance` without discovery and crate a `Provider` for it.
+    ///
+    /// These will all use the same genesis config.
+    fn spawn_geth_and_create_provider(
+        chain_id: u64,
+        rpc_port: u16,
+        p2p_port: u16,
+        datadir: Option<PathBuf>,
+        genesis: Option<Genesis>,
+    ) -> (GethInstance, Provider<HttpProvider>) {
+        let geth =
+            Geth::new().port(rpc_port).p2p_port(p2p_port).chain_id(chain_id).disable_discovery();
+
+        let geth = match genesis {
+            Some(genesis) => geth.genesis(genesis),
+            None => geth,
+        };
+
+        let geth = match datadir {
+            Some(dir) => geth.data_dir(dir),
+            None => geth,
+        }
+        .spawn();
+
+        let url = format!("http://localhost:{}", rpc_port);
+        let provider = Provider::try_from(url).unwrap();
+        (geth, provider)
+    }
+
+    #[tokio::test]
+    async fn add_second_geth_peer() {
+        // init each geth directory
+        let dir1 = tempfile::tempdir().unwrap().into_path();
+        let dir2 = tempfile::tempdir().unwrap().into_path();
+
+        // use the default genesis
+        let genesis = utils::Genesis::default();
+
+        // record p2p ports so they can be used to construct enodes
+        let first_port = 30304u16;
+        let second_port = 30305u16;
+
+        let (first_geth, first_peer) = spawn_geth_and_create_provider(
+            1337,
+            8546,
+            first_port,
+            Some(dir1.clone()),
+            Some(genesis.clone()),
+        );
+        let (second_geth, second_peer) = spawn_geth_and_create_provider(
+            1337,
+            8547,
+            second_port,
+            Some(dir2.clone()),
+            Some(genesis.clone()),
+        );
+
+        // get nodeinfo for each geth instance
+        let first_info = first_peer.node_info().await.unwrap();
+        let second_info = second_peer.node_info().await.unwrap();
+
+        // replace the ip in the enode by putting
+        let first_prefix = first_info.enode.split('@').collect::<Vec<&str>>();
+
+        // create enodes for each geth instance using each id and port
+        let first_enode = format!("{}@localhost:{}", first_prefix.first().unwrap(), first_port);
+
+        // add the second geth as a peer for the first
+        let res = second_peer.add_peer(first_enode).await.unwrap();
+        assert!(res);
+
+        // wait for the geth dial loop to start and connect the static nodes
+        sleep(Duration::from_secs(20)).await;
```

Comment:
this takes that long? can we instead poll the node like we poll ganache-cli instaed of hardcoding a wait time? it's gonna screw up our reth integration tests otherwise imo

Reply:
yeah, added a method `wait_to_add_peer` which polls geth's stderr

Reply:
another method would be polling the `admin_peers` endpoint, we might be able to accomplish this using `admin_peerEvents` in the future